### DISCUSSION TITLE
Background-Color

### DISCUSSION THREAD

**Discussion by felix-d1strict** [chillerlan/php-qrcode#121] [Q&A] [answered]
Hello,

is it possible to set a background color other than white as an option?
White instead of transparent is possible as follows: QROptions::$imageTransparent, but I have not found anything else for other colors.

Many greetings,
Felix

**Comment by codemasher** [comment, marked as answer]
Hey,

i think i should add a bit documentation what `QROptions::$imageTransparent` actually does.

- In [`GdImage` mode](https://github.com/chillerlan/php-qrcode/blob/b6777fd58b2275ff108c75abc9fa9767b180a16b/src/Output/QRGdImage.php#L101-L109) (png, gif) it sets `imagecolortransparent()` with `QROptions::$imageTransparencyBG`. The part that's not documented is that `$imageTransparencyBG` also sets the "normal" background color without transparency switch.
- In [SVG mode](https://github.com/chillerlan/php-qrcode/blob/main/src/Output/QRMarkup.php#L211-L213) (as of v5), it won't render the "light" modules, as opacity/transparency can easily be set with css properties.
- It has no effect in the FPDF and Imagick output modules.

**Reply by felix-d1strict** [reply]
Hello codemasher,

Thanks for the detailed reply (and moving it to the discussions (completely missed this tab)).
Yes, documentation on this would certainly be helpful, and not just for me. 😄
> * `$imageTransparencyBG` also sets the "normal" background color without transparency switch.

That's what I was looking for. Thank you very much. :)

Greetings from Berlin,
Felix

**Reply by codemasher** [reply]
I've added the documentation bit to the [corresponding docblocks](https://github.com/chillerlan/php-qrcode/commit/585c328d9e8c8dc79e52cf7242f2e6eb3b4b91af).
Now i wonder if it makes sense to introduce a separate bgcolor option as i can't simply change behavoiur at this point... 🤔

**Reply by felix-d1strict** [reply]
Thank you very much! 👍🏻 

> Now i wonder if it makes sense to introduce a separate bgcolor option as i can't simply change behavoiur at this point... 🤔

I do think that makes sense. That would at least give us a little more room for “individualization”. So I would be in favor of it. Of course, it also works as you described above – but a separate option would be nicer.

**Reply by codemasher** [reply]
Honestly i'm curious how your qrcodes would look like. 
As far as i understand my own code, the transparency would only occur if a module is the same color as the `$imageTransparencyBG` (aka with the default settings, white background and "light" modules).
The "normal" background would otherwise be invisible at all times as soon as you change the colors for the modules, so it would only be visible if you would draw the modules with an opacity, which is not possible with the default implementation. 
Is there something i've overlooked?

**Comment by robsonvitor** [comment]
Hello,

I have a background that is blue. When I generate the QR Code, the generated image overlays the blue background, leaving the white background and the code in black. How should I do so that the code is generated in white and with a transparent background? I'm using SVG. I tried the options below, but it still ignores it. Furthermore, this QR Code with the background image is saved in a PDF via PHP. 

Any ideas on how to resolve this?

$options = new QROptions;
$options->addQuietzone = false;
$options->imageTransparent = true;

**Reply by codemasher** [reply]
Hi! The [`imageTransparent` option](https://php-qrcode.readthedocs.io/en/main/Usage/Configuration-settings.html#imagetransparent) has no effect for SVG. What you're looking for is the option to [omit the light modules `drawlightmodules`](https://php-qrcode.readthedocs.io/en/main/Usage/Configuration-settings.html#drawlightmodules) - set this to false and you'll see the background shine through as SVG is transparent by default.

**Comment by robsonvitor** [comment]
Very good, @codemasher! Thank you very much for your recommendation, it worked perfectly. The last question is what is the parameter for the QR code to be drawn in white? By default it is black, but as the background color is dark blue, I believe white would be better for reading the QR code.

**Reply by codemasher** [reply]
You can style the SVG via CSS, please check out [the examples](https://github.com/chillerlan/php-qrcode/blob/a301af5b7faa7fcd690b29f38b283616a7d93b89/examples/svg.php).

**Reply by robsonvitor** [reply]
Thank you very much, it worked. However, only when displaying the page. When I generate the PDF, it remains black. Is there any parameter that I must pass to render it in the PDF? Thanks

**Reply by codemasher** [reply]
You can try using the `fill` attributes on the paths. For this you need to set the option [`svgUseFillAttributes`](https://php-qrcode.readthedocs.io/en/main/Usage/Configuration-settings.html#svgusefillattributes) and set the [`modulevalues`](https://php-qrcode.readthedocs.io/en/main/Usage/Configuration-settings.html#modulevalues) option similar as shown in the [imagick example over here](https://github.com/chillerlan/php-qrcode/blob/main/examples/imagick.php). If you have set the [`connectPaths` option](https://php-qrcode.readthedocs.io/en/main/Usage/Configuration-settings.html#connectpaths), you only need to specify the `M_DATA_*` values

**Reply by codemasher** [reply]
Wait a second, i just realized that `svgUseFillAttributes` is set to true by default, which might interfere with other settinngs previously mentioned. Further, it depends on how the conversion to PDF takes place. The better option might be to Convert the SVG to a raster image before embedding it into the PDF, similar to the [imagick conversion example](https://github.com/chillerlan/php-qrcode/blob/a301af5b7faa7fcd690b29f38b283616a7d93b89/examples/imagickConvertSVGtoPNG.php).

**Reply by robsonvitor** [reply]
Thank you very much for all the support. I'll try a little more here, I believe that DOMPDF is making things difficult.

**Reply by codemasher** [reply]
Yea that's possible. Working the DOM outside of a browser engine is not a trivial task and you might be better off with a conversion to a raster image in that case like i suggested. In any case: good luck!
